consensus/polybft/proposer_calculator_test.go

```diff
@@ -442,27 +409,103 @@ func TestProposerCalculator_UpdatesForNewValidatorSet(t *testing.T) {
-	require.NoError(t, err)
+	snapshot := NewProposerSnapshot(0, vs.Accounts())
+	pc := NewProposerCalculator(hclog.NewNullLogger())
 
-	_, err = pc.CalcProposer(1)
+	_, err = pc.CalcProposer(snapshot, 1, 0)
 	require.NoError(t, err)
 
 	// verify that the capacity and length of validators is the same
-	assert.Equal(t, len(vs.Accounts()), cap(pc.validators))
+	assert.Equal(t, len(vs.Accounts()), cap(snapshot.Validators))
 	// verify that validator priorities are centered
-	valsCount := int64(len(pc.validators))
+	valsCount := int64(len(snapshot.Validators))
 
 	sum := int64(0)
-	for _, val := range pc.validators {
+	for _, val := range snapshot.Validators {
 		// mind overflow
 		sum = safeAddClip(sum, val.ProposerPriority)
 	}
 
 	assert.True(t, sum < valsCount && sum > -valsCount,
 		"expected total priority in (-%d, %d). Got %d", valsCount, valsCount, sum)
 	// verify that priorities are scaled
-	dist := computeMaxMinPriorityDiff(pc.validators)
+	dist := computeMaxMinPriorityDiff(snapshot.Validators)
 	assert.True(t, dist <= priorityWindowSizeFactor*vs.totalVotingPower,
 		"expected priority distance < %d. Got %d", priorityWindowSizeFactor*vs.totalVotingPower, dist)
 }
+
+func TestProposerCalculator_GetLatestProposer(t *testing.T) {
+	const (
+		bestIdx = 5
+		count   = 10
+	)
+
+	t.Parallel()
+
+	validatorSet := newTestValidators(count).getPublicIdentities()
+	snap := NewProposerSnapshot(0, validatorSet)
+	proposerCalculator := NewProposerCalculator(hclog.NewNullLogger())
+
+	snap.Validators[bestIdx].ProposerPriority = 1000000
+
+	// not set
+	_, exists := proposerCalculator.GetLatestProposer(snap, 0, 0)
+	assert.False(t, exists)
+
+	_, err := proposerCalculator.CalcProposer(snap, 0, 0)
+	require.NoError(t, err)
+
+	// wrong round
+	_, exists = proposerCalculator.GetLatestProposer(snap, 1, 0)
+	assert.False(t, exists)
+
+	// wrong height
+	_, exists = proposerCalculator.GetLatestProposer(snap, 0, 1)
+	assert.False(t, exists)
+
+	// ok
+	address, exists := proposerCalculator.GetLatestProposer(snap, 0, 0)
+	assert.True(t, exists)
+
+	assert.Equal(t, validatorSet[bestIdx].Address, address)
+}
+
+func TestProposerCalculator_UpdateValidators(t *testing.T) {
+	const rounds = 7
+
+	desiredPriorities := []int64{-300, -300, 0}
+
+	keys, err := bls.CreateRandomBlsKeys(8)
+	require.NoError(t, err)
+
+	v1 := &ValidatorMetadata{Address: types.Address{0x1}, BlsKey: keys[0].PublicKey(), VotingPower: 100}
+	v2 := &ValidatorMetadata{Address: types.Address{0x2}, BlsKey: keys[1].PublicKey(), VotingPower: 100}
+	v3 := &ValidatorMetadata{Address: types.Address{0x3}, BlsKey: keys[2].PublicKey(), VotingPower: 100}
+	v4 := &ValidatorMetadata{Address: types.Address{0x4}, BlsKey: keys[3].PublicKey(), VotingPower: 100}
+	v5 := &ValidatorMetadata{Address: types.Address{0x5}, BlsKey: keys[4].PublicKey(), VotingPower: 100}
+
+	accountSet := []*ValidatorMetadata{v1, v2, v3, v4, v5}
+	vs, err := NewValidatorSet(accountSet, hclog.NewNullLogger())
```

Comment:
Seems like this could have been shortened if `testValidators` abstraction was used...?

Reply:
Creating random address would lead to inconsistency and unpredictable behavior.